### ISSUE TITLE
Improve runtime/memory of cleanlab.segmentation.filter.find_label_issues()

### ISSUE THREAD

**Issue by jwmueller** [cleanlab/cleanlab#863]
# Background info to look at first

Documentation for this method:  https://docs.cleanlab.ai/master/cleanlab/segmentation/filter.html#cleanlab.segmentation.filter.find_label_issues

Related discussion + workarounds you can use in the meantime:  https://github.com/cleanlab/cleanlab/issues/842

# Objective 

Get this `cleanlab.segmentation.filter.find_label_issues()` code to run successfully on a dataset with many standard-sized images & many classes (say 100k+ images, 100+ classes), on a machine with limited RAM (say your laptop).

You should also verify the correctness of your code by comparing its outputs to the outputs of the current code on smaller datasets (they should always match).

# Root cause 

Ideally the `cleanlab.segmentation.filter.find_label_issues()` code should be able to run in mini-batches of images and thus not require massive RAM.  However, a user tried to run it with a mem-mapped array and hit a RAM bottleneck (which should not exist if it's truly running in mini-batches).

I suspect the bottleneck is in this line of code:
https://github.com/cleanlab/cleanlab/blob/b2c28d29172667bc04499087442795e21a33d13a/cleanlab/segmentation/filter.py#L124

where either `flatten()`, `np.moveaxis`, or `reshape()` are causing the memmap array to be loaded entirely into memory all at once and thus causing RAM error.

# Ideas for fixes

Here is one idea for a potential fix, but many others are likely possible: 

The  call to `find_label_issues_batched()` in the original `segmentation.filter.find_label_issues()` code can be replaced with its source `label_inspector.py` code. Note that the internal code used in  `find_label_issues_batched()` is all available to run independently in `label_inspector.py`. Once you have replaced the function to expose it's internals, you can try to modify these (only inside of `cleanlab.segmentation`, do NOT change the code in `label_inspector.py`) appropriately to avoid ever loading the full dataset predictions/labels into memory at once.  One option could be to call `flatten_and_preprocess_masks` within the mini-batches.

Some pseudocode related to this idea:

```
#continuously load images until number of pixels greater than batch size 
	images_to_load=1
	while(number of pixels is less than batch_size)
		images_to_load+=1

#images_to_load is the number of images we can load at a time based on the batch size (default is just 1)

lab = LabelInspector(
The pred_probs shape we can calculate based on the loaded numpy array 
        num_class=pred_probs.shape[1], 
        verbose=verbose,
        n_jobs=n_jobs,
        quality_score_kwargs=quality_score_kwargs,
        num_issue_kwargs=num_issue_kwargs,
    )
n can also be calculated using the loaded array 
    n = len(labels)
    if verbose:
        from tqdm.auto import tqdm

        pbar = tqdm(desc="number of examples processed for estimating thresholds", total=n)
```

For the following code snippet, we change the while loops to be over`images_to_load`. Instead of individual points, we load individual images:

```
labels_batch = flatten_and_process(labels[I:end_index]…)
(Do the same for pred_probs)
    i = 0
    while i < n:
        end_index = i + batch_size
        labels_batch = labels[i:end_index]
        pred_probs_batch = pred_probs[i:end_index, :]
        i = end_index
        lab.update_confident_thresholds(labels_batch, pred_probs_batch)
        if verbose:
            pbar.update(batch_size)

    # Next evaluate the quality of the labels (run this on full dataset you want to evaluate):
    if verbose:
        pbar.close()
        pbar = tqdm(desc="number of examples processed for checking labels", total=n)
    i = 0
    while i < n:
        end_index = i + batch_size
        labels_batch = labels[i:end_index]
        pred_probs_batch = pred_probs[i:end_index, :]
        i = end_index
        _ = lab.score_label_quality(labels_batch, pred_probs_batch)
        if verbose:
            pbar.update(batch_size)

    if verbose:
        pbar.close()

    return lab.get_label_issues()
```

TLDR of this idea is to change the usage of `find_label_issues_batched()` within `cleanlab.segmentation.filter.find_label_issues()`, to iterate over images instead of points and be sure to call `flatten_and_preprocess` internally within the mini-batches to avoid loading the full dataset predictions/labels all at once.

**Comment by jwmueller** [comment]
@MrAnayDongre has started working on a older version of this Issue here:
https://github.com/cleanlab/cleanlab/pull/858

**Comment by kylegallatin** [comment]
Hey @jwmueller, interested in taking this on! Have a few questions for clarification (which will likely be somewhat naive 😄):

> I suspect the bottleneck is in this line of code
Any chance that `flatten_and_preprocess_masks()` could be changed to a chunk-based approach or do you think that would still require loading entire mem-mapped arrays into memory (unless converted to a generator or something). I did actually run some reallllly naive memory profiling tests with this approach using the unit tests, but didn't seem super promising.

> can be replaced with its source `label_inspector.py` code
Just wanted to confirm that this actually refers to the `LabelInspector` class.

> TLDR of this idea is to change the usage of find_label_issues_batched() within cleanlab.segmentation.filter.find_label_issues(), to iterate over images instead of points and be sure to call flatten_and_preprocess internally within the mini-batches to avoid loading the full dataset predictions/labels all at once
Very cool! So:
- replace `find_label_issues_batched` with its source code
- move `flatten_and_preprocess_masks` into that newly copy/pasted source code
- update/expound on the pseudocode provided so that that new code operates at the image level

Final Q (I swear) any recommendations on testing? I was just running the `test_segmentation.py::test_find_label_issues` unit test with some profiling by itself - but if we want to make sure this implementation _specifically_ works for mem-mapped arrays I can come up with something else.

**Comment by jwmueller** [comment]
@kylegallatin Questions are expected! The issue spec could certainly use some clarifications about various points. Overall I think you have a good plan with your 3 bullet points, that is exactly what we were thinking when opening this issue.  

Re your Qs:

1. Yes the `label_inspector.py` code is actually referring to [LabelInspector](https://github.com/cleanlab/cleanlab/blob/ccb4a90e38f6734aa4a12de0f6fd297688dc2ba0/cleanlab/experimental/label_issues_batched.py#L244C7-L244C21) code.

2. You're free to change `flatten_and_preprocess_masks()` however you see fit as long as the results remain the same as the current approach that is internally using `find_label_issues_batched()`.

3. Re unit tests: The main test is to ensure the results from your new code match the current code. 

The second unit test is to verify your new code runs with small batch sizes and a mem-mapped array. So yes you'll want to add another test for that.  You can see how to create a mem-mapped array and load one in the docstring / example-script of [find_label_issues_batched()](https://docs.cleanlab.ai/stable/cleanlab/experimental/label_issues_batched.html#cleanlab.experimental.label_issues_batched.find_label_issues_batched). Maybe the easiest is to just save/load the same array from the current unit test as such a mem-mapped array.

A final test that does NOT need to be unit test (just a one-time script you can comment in PR description) is to check your new code uses less memory when running with small batch-size and a mem-mapped array. To reliably verify this, you'll need a sufficiently large mem-mapped array but you can probably get away with just duplicating the dataset from the existing unit test multiple times. 

Let me know if you have additional questions

**Comment by kylegallatin** [comment]
Is `flatten_and_preprocess_masks()` intended to work with `zarr` arrays? The `Array` object returned by `zarr.convenience.open(YOURFILE.zarr, mode="r")` doesn't seem to have `flatten()` or `reshape()` methods.

Started the process [here](https://github.com/cleanlab/cleanlab/compare/master...kylegallatin:cleanlab:issue-873-improve-mem-find-label-issues?expand=1)

**Comment by jwmueller** [comment]
> Is `flatten_and_preprocess_masks()` intended to work with zarr arrays?

No it's ok to make it just work for saved numpy arrays. We just want this to scale to big datasets for at least one storage-type. Let's just be clear about the requirements on the user's part, i.e. docstring of `segmentation.filter.find_label_issues` should say something like: 

> To scale this to large datasets, you should load/format  your np arrays as follows: ...

**Comment by kylegallatin** [comment]
Initial implementation seems to reduce memory usage for large memmapped arrays by a decent chunk: https://github.com/cleanlab/cleanlab/pull/885

**Comment by kylegallatin** [comment]
Hey @jwmueller, apologies for some delay! I was sick a good chunk of the past week. Updates are as follows:
- I can confirm the new approach uses less memory than the old approach
- I _cannot_ seem to confirm that lower batch sizes yield less memory usage
    -  Going as low as printing the memory per object, it looks as if the memory used by `LabelInspector` makes up the largest chunk and is fairly invariable 
    -  Tried deliberately `del`ing or `gc`ing other objects but unfortunately this didn't seem to alter the results in a meaningful way
    - As a result, it's difficult to think of a long-standing, meaningful unit test involving memory at this time
I'll fix up the build for now, but let me know where you want to take this for the time being with that in mind!

**Comment by jwmueller** [comment]
> Hey @jwmueller, apologies for some delay! I was sick a good chunk of the past week. Updates are as follows:
> 
> * I can confirm the new approach uses less memory than the old approach
> * I _cannot_ seem to confirm that lower batch sizes yield less memory usage
>   
>   * Going as low as printing the memory per object, it looks as if the memory used by `LabelInspector` makes up the largest chunk and is fairly invariable
>   * Tried deliberately `del`ing or `gc`ing other objects but unfortunately this didn't seem to alter the results in a meaningful way
> 
> As a result, it's difficult to think of a long-standing, meaningful unit test involving memory at this time I'll fix up the build for now, but let me know where you want to take this for the time being with that in mind!

Hope you are feeling better now!  No worries about unit-testing the memory, I can see that it could get complicated.  Instead how about just adding one new unit test that additionally verifies the results are identical regardless if the batch size is: full dataset, half the dataset, or 1 (can just use the same dataset already being used for unit-testing the current segmentation code).

Once your code is finalized, please do confirm its results exactly match the current cleanlab code, thanks!

**Comment by kylegallatin** [comment]
PR is open.

**Comment by jwmueller** [comment]
@kylegallatin great stuff, our team is now reviewing your code. stay tuned!

**Comment by kylegallatin** [comment]
Great to hear! Only possible because of your support.

**Comment by jwmueller** [comment]
Completed in: https://github.com/cleanlab/cleanlab/pull/885
Awesome work by @kylegallatin !
